plugin/cache/handler.go

```diff
@@ -27,18 +27,33 @@ func (c *Cache) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg)
 	server := metrics.WithServer(ctx)
 
 	i, found := c.get(now, state, server)
-	if i != nil && found {
-		resp := i.toMsg(r, now)
-		w.WriteMsg(resp)
-
-		if c.shouldPrefetch(i, now) {
-			go c.doPrefetch(ctx, state, server, i, now)
+	if i == nil || !found {
+		if c.serveExpired {
+			i, found = c.get(now.Add(-c.expiredUpTo), state, server)
+		}
+		if !found {
+			crr := &ResponseWriter{ResponseWriter: w, Cache: c, state: state, server: server}
+			return plugin.NextOrFailure(c.Name(), c.Next, ctx, crr, r)
 		}
-		return dns.RcodeSuccess, nil
+		servedExpired.WithLabelValues(server).Inc()
+		// Adjust the time to get a 0 TTL in the reply we got from a stale item.
+		t := i.ttl(now)
+		if t < 0 {
+			now = now.Add(time.Duration(t) * time.Second)
+		}
+		go func() {
+			r := r.Copy()
+			crr := &ResponseWriter{Cache: c, state: state, server: server, prefetch: true, remoteAddr: w.LocalAddr()}
```

Comment:
IIUC, this go routine is the "ends up in the cache later on" bit.
Will this cause CoreDNS to send a DNS response packet to itself?

Reply:
No, nothing is written to the response when prefetch is true.
The LocalAddr is arbitrary and is only needed to avoid a nullref in one of the ResponseWriter methods. Once a backend returns a cacheable response, it is cached but never written.